### PR TITLE
Feat/add abstract component methods

### PR DESCRIPTION
**Pull request by fhdhsni**
fixes: #9

### CODE REVIEW

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java [hunk 1]
```diff
@@ -15,5 +15,42 @@
...
+        add(components);
+    }
+
+    @Override
+    public void moveComponentsFrom(ComponentContainer source) {
+        Component[] components = source.getElement().getChildren()
+                .map(Element::getComponent)
+                .flatMap(maybeComponent -> maybeComponent.map(Stream::of)
+                        .orElseGet(Stream::empty))
+                .toArray(Component[]::new);
```
Comment (pleku) [inline vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java:42]: Just noting that none of the methods are unit tested, which is OK for now except in my opinion for this method. I would like to at least see a test for validating that any children that are not components are not failing - at the moment like it works like that, but without tests, door is open for regressions.

**Review by pleku — CHANGES_REQUESTED** [review]

File: vaadin-legacy-components-flow-integration-tests/src/main/java/com/vaadin/legacy/ui/examples/layout/layouttester/BaseAddReplaceMove.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        addBtn.setId(idPrefix + "add-btn");
+        final NativeButton moveBtn = new NativeButton("Test move");
+        moveBtn.setId(idPrefix + "move-btn");
+        final NativeButton removeBtn = new NativeButton("Test remove");
+        removeBtn.setId(idPrefix + "remove-btn");
+
+        addBtn.addClickListener(
+                event -> layout.addComponent(new Paragraph("six")));
+
+        // TODO
```
[the remark this answers is not in the document]
Comment (fhdhsni) [reply vaadin-legacy-components-flow-integration-tests/src/main/java/com/vaadin/legacy/ui/examples/layout/layouttester/BaseAddReplaceMove.java]: done

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java [hunk 3]
```diff
@@ -15,5 +15,42 @@
  * #L%
  */
 
-public abstract class AbstractComponentContainer extends AbstractComponent {
+import com.vaadin.flow.component.Component;
+import com.vaadin.flow.dom.Element;
+
+import java.util.stream.Stream;
+
+public abstract class AbstractComponentContainer extends AbstractComponent
```
[the remark this answers is not in the document]
Comment (fhdhsni) [reply vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java:35]: I thought the docs will be inherited from the interface. Copy and pasted for now and mentioned that this is a legacy component

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractOrderedLayout.java [hunk 4]
```diff
@@ -15,8 +15,21 @@
...
+import com.vaadin.flow.component.HasSize;
+import com.vaadin.flow.component.HasStyle;
 
 public abstract class AbstractOrderedLayout extends AbstractLayout
-        implements HasComponents, HasSize, HasStyle {
+        implements HasSize, HasStyle {
+    @Override
+    public void replaceComponent(Component oldComponent,
+            Component newComponent) {
+        // TODO
```
[the remark this answers is not in the document]
Comment (fhdhsni) [reply vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractOrderedLayout.java]: done

File: vaadin-legacy-components-testbench/src/main/java/com/vaadin/legacy/ui/testbench/elements/VerticalLayoutElement.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.vaadin.legacy.ui.testbench.elements;
```
[the remark this answers is not in the document]
Comment (fhdhsni) [reply vaadin-legacy-components-testbench/src/main/java/com/vaadin/legacy/ui/testbench/elements/VerticalLayoutElement.java]: changed to package to `com.vaadin.testbench.elements;`

File: vaadin-legacy-components-testbench/src/main/java/com/vaadin/legacy/ui/testbench/elements/AbstractComponentContainerElement.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package com.vaadin.legacy.ui.testbench.elements;
+
+import org.openqa.selenium.WebElement;
+
+import java.util.List;
+
+public class AbstractComponentContainerElement
+        extends AbstractComponentElement {
+    public List<WebElement> getChildren() {
```
[the remark this answers is not in the document]
Comment (fhdhsni) [reply vaadin-legacy-components-testbench/src/main/java/com/vaadin/legacy/ui/testbench/elements/AbstractComponentContainerElement.java]: done

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java (hunk 1 above)
Comment (fhdhsni) [reply vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java:74]: added a test with a `source` that has an `Element` as a child without a corresponding component

**Review by pleku — APPROVED** [review]
LGTM even though there are extra javadocs on methods (see comment https://github.com/vaadin/flow-legacy-components/pull/23#discussion_r702866235)

File: vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (pleku) [reply vaadin-legacy-components-flow/src/main/java/com/vaadin/legacy/ui/AbstractComponentContainer.java:35]: The javadocs will be inherited for methods when overriding them from parent class.

It is not the same when creating a new class; even though it implements an interface or extends another class, it needs its own javadocs. My comment was about the class level javadoc, not the method ones.

File: vaadin-legacy-components-flow/pom.xml [hunk 7]
```diff
@@ -34,6 +34,10 @@
             <artifactId>flow-data</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>com.vaadin</groupId>
+            <artifactId>flow-html-components</artifactId>
```
Comment (pleku) [inline vaadin-legacy-components-flow/pom.xml:39]: At the moment only needed for tests ? 
```suggestion
            <artifactId>flow-html-components</artifactId>
            <scope>test</scope>
```

**Review by pleku — CHANGES_REQUESTED** [review]

**Review by pleku — APPROVED** [review]

**Pull request merged**